flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/SavepointDeepCopyTest.java

```diff
@@ -0,0 +1,212 @@
+
+import org.apache.flink.api.common.state.ValueState;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.flink.api.common.typeinfo.TypeHint;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.contrib.streaming.state.RocksDBStateBackend;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.state.StateBackend;
+import org.apache.flink.runtime.state.filesystem.FsStateBackend;
+import org.apache.flink.state.api.functions.KeyedStateBootstrapFunction;
+import org.apache.flink.state.api.functions.KeyedStateReaderFunction;
+import org.apache.flink.test.util.AbstractTestBase;
+import org.apache.flink.util.AbstractID;
+import org.apache.flink.util.Collector;
+
+import org.apache.commons.lang3.RandomStringUtils;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.Matchers.everyItem;
+import static org.hamcrest.Matchers.isIn;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Test the savepoint deep copy.
+ */
+@RunWith(value = Parameterized.class)
+public class SavepointDeepCopyTest extends AbstractTestBase {
+
+	private static final String TEXT = "The quick brown fox jumps over the lazy dog";
+	private static final String RANDOM_VALUE = RandomStringUtils.randomAlphanumeric(120);
+
+	private final StateBackend backend;
+
+	public SavepointDeepCopyTest(StateBackend backend) throws Exception {
+		this.backend = backend;
+		//reset the cluster so we can change the state backend
+		miniClusterResource.after();
+		miniClusterResource.before();
+	}
+
+	@Parameterized.Parameters(name = "State Backend: {0}")
+	public static Collection<StateBackend> data() {
+		// set the threshold to 1024 bytes to allow generate additional state data files with a small state
+		int fileStateSizeThreshold = 1024;
+		return Arrays.asList(
+			new FsStateBackend(new Path("file:///tmp").toUri(), fileStateSizeThreshold),
+			new RocksDBStateBackend(
+				(StateBackend) new FsStateBackend(new Path("file:///tmp").toUri(), fileStateSizeThreshold)
+			)
+		);
+	}
+
+	/**
+	 * To bootstrapper a savepoint for testing.
+	 */
+	static class WordMapBootstrapper extends KeyedStateBootstrapFunction<String, String> {
+		private ValueState<Tuple2<String, String>> state;
+
+		@Override
+		public void open(Configuration parameters) {
+			ValueStateDescriptor<Tuple2<String, String>> descriptor = new ValueStateDescriptor<>(
+				"state",
+				TypeInformation.of(new TypeHint<Tuple2<String, String>>(){}));
```

Comment:
fyi, there's an easier way to do this. 
```suggestion
				Types.Tuple(Types.String, Types.String)
```

Reply:
Thanks for the hint!